lib/battle/handler/BattlefieldHandler.cpp

```diff
@@ -0,0 +1,61 @@
+/*
+ * BattlefieldHandler.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include "StdInc.h"
+#include "BattlefieldHandler.h"
+#include "CModHandler.h"
+#include "battle/obstacle/ObstacleJson.h"
+
+BattlefieldHandler::BattlefieldHandler()
+{
+	const JsonNode obstacleConfig(JsonNode(ResourceID("config/obstacles.json")));
+
+	for(auto i : obstacleConfig.Struct())
+	{
+		obstacleConfigs.push_back(std::make_shared<ObstacleJson>(i.second));
+	}
+}
+
+BattlefieldHandler::~BattlefieldHandler()
+{
+
+}
+
+void BattlefieldHandler::loadObject(std::string scope, std::string name, const JsonNode & data)
+{
+	auto object = std::make_shared<ObstacleJson>(data);
+	obstacleConfigs.push_back(object);
+}
+
+void BattlefieldHandler::loadObject(std::string scope, std::string name, const JsonNode & data, size_t index)
+{
+	auto object = std::make_shared<ObstacleJson>(data);
+	obstacleConfigs[index] = object;
+}
+
+std::vector<JsonNode> BattlefieldHandler::loadLegacyData(size_t dataSize)
+{
+	std::vector<JsonNode> data;
```

Comment:
* Do not use legacy data if you actually do have it.
* even if you use legacy data API, `loadLegacyData` must return vector of individual entities, but not 1-element vector with entire configuration